File: src/main/java/cn/enaium/joe/gui/panel/menu/MappingMenu.java
/*
 * Copyright 2022 Enaium
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.enaium.joe.gui.panel.menu;

import cn.enaium.joe.JavaOctetEditor;
import cn.enaium.joe.task.RemappingTask;
import cn.enaium.joe.util.JFileChooserUtil;
import cn.enaium.joe.util.LangUtil;
import net.fabricmc.mappingio.format.MappingFormat;

import javax.swing.*;
import javax.swing.filechooser.FileFilter;
import java.io.File;

/**
 * @author Enaium
 * @since 1.0.0
 */
public class MappingMenu extends JMenu {
    public MappingMenu() {
        super(LangUtil.i18n("menu.mapping"));
        for (MappingFormat value : MappingFormat.values()) {
            add(new JMenuItem(value.name) {{
                addActionListener(e -> {
                    File show = JFileChooserUtil.show(JFileChooserUtil.Type.OPEN, new FileFilter() {
                        @Override
                        public boolean accept(File f) {
                            return f.getName().matches(".+(txt|map|mapping|pro|srg|tsrg|tiny|tiyv2)") || f.isDirectory();
                        }

                        @Override
                        public String getDescription() {
                            return "Mapping file(*.txt,*.map,*.mapping,*.pro,*.srg)";
                        }
                    });
                    if (JavaOctetEditor.getInstance().getJar() != null && show != null) {
                        JavaOctetEditor.getInstance().task.submit(new RemappingTask(show, value));
                    }
                });
            }});
        }
    }
}

File: ui/src/main/kotlin/cn/enaium/joe/ui/control/menu/MappingMenu.kt
/*
 * Copyright 2022 Enaium
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.enaium.joe.ui.control.menu

import cn.enaium.joe.core.task.RemappingTask
import cn.enaium.joe.ui.Instance.jar
import cn.enaium.joe.ui.JavaOctetEditor
import cn.enaium.joe.ui.JavaOctetEditor.Companion.event
import cn.enaium.joe.ui.JavaOctetEditor.Companion.task
import cn.enaium.joe.ui.event.LoadJar
import cn.enaium.joe.ui.util.i18n
import javafx.scene.control.Menu
import javafx.scene.control.MenuItem
import javafx.stage.FileChooser
import net.fabricmc.mappingio.format.MappingFormat
import java.io.File

/**
 * @author Enaium
 * @since 2.0.0
 */
class MappingMenu : Menu(i18n("menu.mapping")) {
    init {
        MappingFormat.values().forEach { value ->
            items.add(MenuItem(value.name).apply {
                setOnAction {
                    val fileChooser = FileChooser()
                    fileChooser.title = "Open Mapping File"
                    fileChooser.initialDirectory = File(System.getProperty("user.home"))
                    fileChooser.extensionFilters.addAll(
                        FileChooser.ExtensionFilter(
                            "Mapping file(*.txt,*.map,*.mapping,*.pro,*.srg)",
                            "*.txt",
                            "*.map",
                            "*.mapping",
                            "*.pro",
                            "*.srg",
                            "*.tsrg",
                            "*.tiny",
                            "*.tiyv2"
                        )
                    )
                    val showOpenDialog = fileChooser.showOpenDialog(JavaOctetEditor.stage)
                    if (jar != null && showOpenDialog != null) {
                        task.submit(RemappingTask(jar, showOpenDialog, value)).thenAccept {
                            event.call(LoadJar(it))
                        }
                    }
                }
            })
        }
    }
}